datameta/api/services.py

```diff
@@ -0,0 +1,183 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from pyramid.view import view_config
+from pyramid.request import Request
+from pyramid.httpexceptions import HTTPNoContent, HTTPUnauthorized, HTTPNotFound, HTTPForbidden
+from sqlalchemy.exc import IntegrityError
+from dataclasses import dataclass
+from typing import List
+
+from . import DataHolderBase
+from .. import security, errors, resource, siteid
+from ..errors import get_validation_error
+from ..security import authz
+from ..models import Service, ServiceExecution, User
+from ..resource import get_identifier, resource_by_id
+
+@dataclass
+class ServiceResponse(DataHolderBase):
+    id: dict
+    name: str
+    user_ids: list
+
+@view_config(
+    route_name="services",
+    renderer='json', 
+    request_method="POST", 
+    openapi=True
+)
+def post(request: Request):
+    """POST a new service"""
+    # Revalidate User
+    auth_user = security.revalidate_user(request)
+
+    # Check if the user is authorized to create a service
+    if not authz.create_service(auth_user):
+        raise HTTPForbidden()
+
+    name = request.openapi_validated.body.get("name")
+
+    db = request.dbsession
+
+    # Try to create a new service, catch Integrity Error, if a service with that name already exists
+    try:
+        service = Service(
+            name = name,
+            site_id = siteid.generate(request, Service))
+        db.add(service)
+        db.flush()
+    except IntegrityError:
+        raise get_validation_error(["A service with that name already exists."])    
+
+    # Return for the new service the ID, name and all associated Users
+    services = []
+
+    users = []
+    for user in service.users:
+        users.append(get_identifier(user))
+
+    services.append(ServiceResponse(
+        id = resource.get_identifier(service),
+        name = service.name,
+        user_ids = users
+    ))
+
+    return services
+
+@view_config(
+    route_name="services",
+    renderer='json', 
+    request_method="GET", 
+    openapi=True
+)
+def get(request: Request) -> List[ServiceResponse]:
+    """Get a representation of a service"""
+    # Revalidate User
+    auth_user = security.revalidate_user(request)
+
+    # Check if the user is authorized to view services
+    if not authz.view_services(auth_user):
+        raise HTTPForbidden()
+
+    db = request.dbsession
+
+    # Get for all services the ID, name and all associated Users
+    services = []
+
+    for service in db.query(Service):
```

Comment:
Save additional queries by immediately joining the users:
```suggestion
    for service in db.query(Service).options(joinedload(Service.users)):
```